### PR TITLE
[BUG] Fix mne.viz.Brain.add_volume_labels matrix ordering bug

### PR DESCRIPTION
**Pull request by alexrockhill**
Fixes https://github.com/mne-tools/mne-python/issues/11728

### CODE REVIEW

**Comment by alexrockhill** [comment]
This works for me now

```
import os.path as op
import mne
mne.viz.set_3d_backend('pyvistaqt') # GUI requires pyvista backend
mne.viz.set_3d_options(depth_peeling=False, antialias=False, multi_samples=1) # for bugs of pure black plot of 3D viz

fsaverage_dir = op.dirname(mne.datasets.fetch_fsaverage())
mne.datasets.fetch_hcp_mmp_parcellation()

brain = mne.viz.Brain(
    "fsaverage",
    subjects_dir=fsaverage_dir,
    cortex='low_contrast', alpha=0.2, background='white'
)

brain.add_volume_labels(aseg="aparc+aseg")
```

File: doc/changes/latest.inc [hunk 1]
```diff
@@ -28,6 +28,7 @@ Enhancements
 
 Bugs
 ~~~~
+- Fix bug where :meth:`mne.viz.Brain.add_volume_labels` used an incorrect orientation (:gh:`11730` by `Alex Rockhill`_)
```
Comment (larsoner) [inline doc/changes/latest.inc:31]: Can you move this to after the `:newcontrib:`? We try to keep those at the top

File: mne/viz/_brain/_brain.py [hunk 2]
```diff
@@ -2644,6 +2644,14 @@ def add_volume_labels(
...
         mult = 1e-3 if self._units == "m" else 1
         vox_mri_t[:3] *= mult
+        ornt = nib.orientations.axcodes2ornt(
+            nib.orientations.aff2axcodes(aseg.affine)
+        ).astype(int)
+        ras_ornt = nib.orientations.axcodes2ornt("RAS")
+        ornt_trans = nib.orientations.ornt_transform(ornt, ras_ornt)
+        aseg_data = nib.orientations.apply_orientation(aseg_data, ornt_trans)
+        aff_trans = nib.orientations.inv_ornt_aff(ornt_trans, aseg.shape)
+        vox_mri_t = np.dot(vox_mri_t, aff_trans)
```
Comment (larsoner) [inline mne/viz/_brain/_brain.py:2654]: I am confused why this is necessary -- the `vox_mri_t` should already take care of putting everything in MRI surface RAS coordinates for us. Also, there are a number of places we use `vox_mri_t` and don't use any `axcodes2ornt` stuff, so it's weird this is the one place we need to do it.

Comment (larsoner) [reply mne/viz/_brain/_brain.py:2654]: ... to me it suggests maybe there is some bug with how `_marching_cubes` works or something, rather than some adjustment we need to make to `vox_mri_t`.

I would investigate deeper by using `freeview` and some specific voxel `i, j, k` in some known subvolume and what their viewer says for MRI surface RAS `x, y, z`, confirm that this is just `(i, j, k) @ vox_mri_t` (it had better be!), then inspect the outputs at various points of this function for that label. It should tell us where the problem is...

Comment (alexrockhill) [reply mne/viz/_brain/_brain.py]: So it looks like it was caused by the fortran order check in `_marching_cubes`. I think the solution I just pushed works, here is one issue where the order needed to be fortran https://github.com/mne-tools/mne-python/issues/10860 which lead to the order being auto-detected with `"A"`. Checking to be sure that that works now too.

**Comment by alexrockhill** [comment]
Huh so oddly, in the tumor case, the `np.array.flags.c_contiguous` is `True` and `f_contiguous` is `False` but `order="F"` makes the tumor look normal and in the case of this issue everything is opposite. That would suggest something like `order="F" if image.flags.c_contiguous else "C"` which looks very confusing but I think would work.

**Comment by alexrockhill** [comment]
```
import os.path as op
import mne
mne.viz.set_3d_backend('pyvistaqt') # GUI requires pyvista backend
mne.viz.set_3d_options(depth_peeling=False, antialias=False, multi_samples=1) # for bugs of pure black plot of 3D viz
 
fsaverage_dir = op.dirname(mne.datasets.fetch_fsaverage())
mne.datasets.fetch_hcp_mmp_parcellation()
  
brain = mne.viz.Brain(
     "fsaverage",
      subjects_dir=fsaverage_dir,
      cortex='low_contrast', alpha=0.2, background='white'
) 
brain.add_volume_labels(aseg="aparc+aseg")
```

```
import nrrd
import mne
from mne.surface import _marching_cubes

data, header = nrrd.read('tumor.nrrd', index_order='F')
verts, faces = _marching_cubes(data.T, [1], smooth=0)[0]
fig = mne.viz.create_3d_figure((600, 600), scene=False)
fig.mesh(*verts.T, faces, 'red')
fig.show()
```

Both of these work now using the tumor.nrrd file from here: https://github.com/mne-tools/mne-python/issues/10860. It's not ideal that it's opposite of what you'd expect but I'm not really sure if there's a workaround that's better other than just documenting it with a verbose comment.

**Comment by alexrockhill** [comment]
Okay, after much debugging, part of the issue was with `image = image.copy()` which made a shallow copy that changed the index ordering from Fortran to C without actually changing the index ordering. I checked and this version works for the above two MWE as well as the sEEG example and https://mne.tools/dev/auto_tutorials/inverse/60_visualize_stc.html#sphx-glr-auto-tutorials-inverse-60-visualize-stc-py which are the two uses of `add_volume_labels` and the latter looks like it's currently broken on `main` so this will fix that. Should be good to go.

**Comment by alexrockhill** [comment]
Oh also I in doing this I read into the `vtk` functions and it says that if you delete the reference image in a `numpy_to_vtk` you can cause a segfault so I didn't do that. We might do that other places so maybe a good FYI @larsoner.

```
numpy_to_vtk(num_array, deep=0, array_type=None)
    Converts a real numpy Array to a VTK array object.
    
    This function only works for real arrays.
    Complex arrays are NOT handled.  It also works for multi-component
    arrays.  However, only 1, and 2 dimensional arrays are supported.
    This function is very efficient, so large arrays should not be a
    problem.
    
    If the second argument is set to 1, the array is deep-copied from
    from numpy. This is not as efficient as the default behavior
    (shallow copy) and uses more memory but detaches the two arrays
    such that the numpy array can be released.
    
    WARNING: You must maintain a reference to the passed numpy array, if
    the numpy data is gc'd and VTK will point to garbage which will in
    the best case give you a segfault.
    
    Parameters:
    
    num_array
      a 1D or 2D, real numpy array.
```

File: mne/surface.py [hunk 3]
```diff
@@ -1917,20 +1917,20 @@ def _marching_cubes(image, level, smooth=0, fill_hole_size=None, use_flying_edge
             f"{level.size} elements"
         )
 
+    image_shape = image.shape
+    order = "C" if image.flags.c_contiguous else "F"
+    image = deepcopy(image)  # don't modify original
```
Comment (larsoner) [inline mne/surface.py:1922]: Better than this I think would be to do `image = np.ascontiguousarray(image)`, it guarantees C-contiguity so you don't need any more conditional. Also arrays can be neither C nor F contiguous (e.g., if there is a non-unity stride) so the current code would be broken for such arrays

File: mne/surface.py [hunk 4]
```diff
@@ -1978,8 +1978,11 @@ def _marching_cubes(image, level, smooth=0, fill_hole_size=None, use_flying_edge
         polydata = geometry.GetOutput()
         rr = vtk_to_numpy(polydata.GetPoints().GetData())
         tris = vtk_to_numpy(polydata.GetPolys().GetConnectivityArray()).reshape(-1, 3)
-        rr = np.ascontiguousarray(rr[:, ::-1])
-        tris = np.ascontiguousarray(tris[:, ::-1])
+        if order == "C":
```
Comment (larsoner) [inline mne/surface.py:1981]: Then you don't need this conditional and should be able to revert this change (since the `::-1` is the `fliplr` equiv anyway)

File: mne/surface.py (hunk 3 above)
Comment (alexrockhill) [reply mne/surface.py:1922]: I tried that and it didn't work for the tumor test case. Want to give it a quick look?

Comment (alexrockhill) [reply mne/surface.py:1922]: Yep, you need the tumor nrrd file and to pip install pynrrd from the original issue too. If it resembles a blob, then it worked, otherwise you get a cube of dots.

File: mne/surface.py [hunk 5]
```diff
@@ -1917,6 +1917,14 @@ def _marching_cubes(image, level, smooth=0, fill_hole_size=None, use_flying_edge
             f"{level.size} elements"
         )
 
+    # vtkImageData indexes as slice, row, col (Z, Y, X):
+    # https://discourse.vtk.org/t/very-confused-about-imdata-matrix-index-order/6608/2
+    # We can accomplish his by requiring F continuity and raveling later.
+    # We also pass double as passing integer types directly can be problematic!
+    image = np.asfortranarray(image, dtype=float)
+    image_shape = image.shape
+    assert image.flags.f_contiguous
```
Comment (larsoner) [inline mne/surface.py:1926]: @alexrockhill feel free to look and see if this makes sense

Comment (alexrockhill) [reply mne/surface.py:1926]: Yes, that makes sense, I assume it works. I wasn't able to figure out what SetScalars was expecting the matrix order to be, much better job Googling than me. Seems like a good solution.

**Comment by Dr-Chen-Xiaoyu** [comment]
The lastest dev version of MNE correctly plots. 👍👍👍   Thanks a lot ！！！
